File: 2023/aoc2023/day25.py
def run(g, start):
    seen = set([start])
    queue = [start]
    res = {start: None}  # node: prev
    while queue:
        current = queue.pop()
        for n in g[current]:
            if n in seen:
                continue
            seen.add(n)
            queue.append(n)
            res[n] = current
    return res


def getTree(g, node):
    res = run(g, node)
    tree = {}
    for k, v in res.items():
        if v is None:
            continue
        if v not in tree:
            tree[v] = []
        tree[v].append(k)
    return tree


def parseGraph():
    with open("data/day25.txt") as f:
        data = [line.strip().split(": ") for line in f]

    g = {}
    for key, vstr in data:
        vs = vstr.split()
        g[key] = g.get(key, []) + vs
        for v in vs:
            g[v] = g.get(v, []) + [key]
    return g


def childsof(tree, node):
    res = list(tree.get(node, []))
    for cs in tree.get(node, []):
        res += childsof(tree, cs)
    return res


def parentsof(tree, node):
    for k, vs in tree.items():
        if node in vs:
            return [k] + parentsof(tree, k)
    return []


def main():
    g = parseGraph()

    nodes = tuple(g.keys())
    node = nodes[0]

    print(node)
    tree = getTree(g, node)
    for right in tree.values():
        for candidate in right:
            cs = childsof(tree, candidate) + [candidate]
            ps = parentsof(tree, candidate)
            rem = [i for i in nodes if i not in ps and i not in cs]
            for r in rem:
                if len(set(g[r]) & set(cs)) == 0:
                    ps += [r]
                elif len(set(g[r]) & set(ps)) == 0:
                    cs += [r]
            rem = [i for i in nodes if i not in ps and i not in cs]
            if len(rem) == 2:
                for r in rem:
                    if len(set(g[r]) & set(cs)) == 1:
                        ps += [r]
                    else:
                        cs += [r]

                return len(cs) * len(ps)


if __name__ == "__main__":
    print(main())

File: 2023/aoc2023/lib/day25.ex
defmodule Day25 do
  def main do
    rows = read()

    g =
      Enum.reduce(rows, %{}, fn [key, vals], map ->
        vals = String.split(vals)
        map = Map.update(map, key, vals, fn prev -> vals ++ prev end)

        Enum.reduce(vals, map, fn v, map ->
          Map.update(map, v, [key], fn prev -> [key | prev] end)
        end)
      end)

    runkey(g, key)
  end

  def runkey(g, key) do
    Enum.map(g[key],fn v -> runv(g, key, v) end
  end

  def runv(g, left, right) do

  end

  def get_edges(g) do
    Enum.map(g, fn {k, vs} ->
      Enum.zip(Stream.cycle([k]), vs)
      |> Enum.map(fn {x, y} ->
        [x, y]
        |> Enum.sort()
      end)
    end)
    |> Enum.concat()
    |> Enum.uniq()
  end

  def remove_edge(g, left, right) do
    Enum.map(g, fn {k, v} ->
      cond do
        k == left -> {k, v -- [right]}
        k == right -> {k, v -- [left]}
        true -> {k, v}
      end
    end)
    |> Enum.into(%{})
  end

  def run([], _g, seen), do: seen

  def run([current | queue], g, seen) do
    next = Map.get(g, current) |> Enum.filter(fn n -> not MapSet.member?(seen, n) end)

    seen = Enum.reduce(next, seen, fn n, seen -> MapSet.put(seen, n) end)
    run(next ++ queue, g, seen)
  end

  def read() do
    File.stream!("data/day25_test.txt")
    |> Enum.map(fn line ->
      line
      |> String.trim()
      |> String.split(": ")
    end)
  end
end
